packages/material-ui-styles/src/styled/styled.js

```diff
@@ -37,14 +36,10 @@ function styled(Component) {
 
     if (process.env.NODE_ENV !== 'production' && !name) {
       // Provide a better DX outside production.
-      classNamePrefix = getDisplayName(Component);
-      warning(
-        typeof classNamePrefix === 'string',
-        [
-          'Material-UI: the component displayName is invalid. It needs to be a string.',
```

Comment:
Since this is DX only there's no need to warn here. Having components without a displayName is perfectly fine. React guards against undefined display names as well with callbacks.

Reply:
The warning was introduced for https://github.com/mui-org/material-ui/issues/11942#issuecomment-399252238.
What do you think of keeping the warning while ignoring the case when we can't compute it (undefined) (e.g. with [React.memo](https://github.com/mui-org/material-ui/issues/15887#issuecomment-496509131))?

Reply:
You want a warning if e.g. numbers are provided? I don't think that's very useful. Especially considering it ends up as a string anyway.

Reply:
We shouldn't warn when a number is provided, for #11942 (if nothing has changed in the logic since then) providing a number crash with the current logic.

Reply:
So this is all good? 

Specifically #11942 can be solved with types. We really shouldn't do type checking at runtime inconsistently (compared to propTypes where we always do it).

Reply:
Yes, it sounds good to me.